classes/migration/upgrade/v3_5_0/I9707_WeblateUILocales.php

```diff
@@ -15,18 +15,206 @@
 namespace PKP\migration\upgrade\v3_5_0;
 
 use Illuminate\Database\PostgresConnection;
+use Illuminate\Database\Query\Builder;
 use Illuminate\Support\Facades\DB;
 use PKP\install\DowngradeNotSupportedException;
 use PKP\migration\Migration;
 
-class I9707_WeblateUILocales extends Migration
+abstract class I9707_WeblateUILocales extends Migration
 {
+    protected string $CONTEXT_TABLE = '';
+    protected string $CONTEXT_SETTINGS_TABLE = '';
+    protected string $CONTEXT_COLUMN = '';
+
     /**
      * Run the migrations.
      */
     public function up(): void
     {
-        $localesTable = [
+        $affectedLocales = $this->getAffectedLocales();
+
+        // update all locale and primary_locale columns
+        $schemaLocName = (DB::connection() instanceof PostgresConnection) ? 'TABLE_CATALOG' : 'TABLE_SCHEMA';
+        $renameLocale = fn (string $l) => collect(DB::select("SELECT DISTINCT TABLE_NAME FROM INFORMATION_SCHEMA.COLUMNS WHERE COLUMN_NAME = ? AND {$schemaLocName} = ?", [$l, DB::connection()->getDatabaseName()]))
+            ->each(function (\stdClass $sc) use ($l, $affectedLocales) {
+                foreach ($affectedLocales as $uiLocale => $weblateLocale) {
+                    DB::table($sc->TABLE_NAME ?? $sc->table_name)->where($l, '=', $uiLocale)->update([$l => $weblateLocale]);
+                }
+            });
+        $renameLocale('primary_locale');
+        $renameLocale('locale');
+
+        // site supported and installed locales
+        $site = DB::table('site')
+            ->select(['supported_locales', 'installed_locales'])
+            ->first();
+        $this->updateArrayLocale($site->supported_locales, 'site', 'supported_locales');
+        $this->updateArrayLocale($site->installed_locales, 'site', 'installed_locales');
+
+        // users locales
+        $migration = $this;
+        DB::table('users')->chunkById(1000, function ($users) use ($migration) {
+            foreach ($users as $user) {
+                $migration->updateArrayLocale($user->locales, 'users', 'locales', null, 'user_id', $user->user_id);
+            }
+        }, 'user_id');
```

Comment:
As these "array" columns are using JSON, then the `updateArrayLocale()` method could be updated to issue a single query, basically a sequence of self-enclosed replaced (e.g. `locales = REPLACE(REPLACE(locales, '"X"', "A"), '"Y"', "B")`).

Some installations have thousands of users, due to spammers, so anything that avoids looping through the users is worth in my opinion.